### PR TITLE
[PROPOSAL] Add summary, parent and kind as optional fields for tags

### PR DESCRIPTION
**Pull request by lornajane**
After lots of discussion (and some lively issues that have been open for YEARS), I've put together a proposal for enhancing tags in a future version of OpenAPI (targeting 3.2 as the next minor release). Please take a look, identify problems/gaps/mistakes, identify any use cases of your own where this could help, and let's try to put together something great for a future release.

### CODE REVIEW

**Comment by lornajane** [comment]
I used 3.1.0 as a basis so probably will need to update/reformat after a 3.1.1 release, try not to get hung up on the formatting as I'm not making the main PR quite yet.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (handrews) [inline proposals/2024-09-01-Tags-Improvement.md:32]: _Thank you!!!_  We need more of this kind of tool to spec feedback loop.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (handrews) [inline proposals/2024-09-01-Tags-Improvement.md:48]: (Warning: bikeshedding ahead, feel free to ignore). As "kind" has two meanings would "category" be more clear for non-native speakers?  I agree on avoiding "type" as it usually means data type.

**Review by handrews — DISMISSED** [review]
Looks great. One minor thing that need not hold up the proposal so I'll go ahead and approve.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:13]:
```suggestion
|Issues | [1367](https://github.com/OAI/OpenAPI-Specification/issues/1367) |
```

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:30]: "required exactly one tag per operation" - is that an example for (ab)use of tags?

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:36]:
```suggestion
There are several examples "in the wild" of where a better tags implementation would have helped. Here is a selection of publicly-accessible examples to illustrate some of the problems this proposal could help with:
```

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:128]: Can we add the `external` tag?

**Review by ralfhandl — COMMENTED** [review]
+1, minor nits

File: proposals/2024-09-01-Tags-Improvement.md
Comment (handrews) [reply proposals/2024-09-01-Tags-Improvement.md:128]: @ralfhandl I view this text as just giving us the general idea of how the proposal would be implemented.  We can nail down the spec wording and examples when the spec changes are made.

(we should probably figure out what the balance is between the proposals and the specification text, idk, maybe folks want to get the wording right here, in which case I'm happy to defer to that approach)

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [reply proposals/2024-09-01-Tags-Improvement.md:48]: "Category" has its own heap of baggage, I find "kind" less overloaded and easier to understand here.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [reply proposals/2024-09-01-Tags-Improvement.md:128]: Depending on how long this proposal will sit here without being incorporated into a specification version I'd like to avoid the impression that the proposal allows/advertises the use of dangling "parent" values by adding just one more tag object to the example and make it self-contained.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:116]:
```suggestion
{
	"name": "account-updates",
  "summary": "Account Updates",
	"description": "Account update operations",
  "kind": "nav"
},
{
	"name": "partner",
  "summary": "Partner",
	"description": "Operations available to the partners network",
  "parent": "external",
  "kind": "audience"
},
{
	"name": "external",
  "summary": "External",
	"description": "Operations available to external consumers",
  "kind": "audience"
}
```

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [inline proposals/2024-09-01-Tags-Improvement.md:129]:
```suggestion
- name: account-updates
  summary: Account Updates
  description: Account update operations
  kind: nav

- name: partner
  summary: Partner
  description: Operations available to the partners network
  parent: external
  kind: audience

- name: external
  summary: External
  description: Operations available to external consumers
  kind: audience
```

File: proposals/2024-09-01-Tags-Improvement.md
Comment (handrews) [reply proposals/2024-09-01-Tags-Improvement.md:48]: @ralfhandl I'm fine with not using "category" but I am curious as to what baggage it has in this context?

File: proposals/2024-09-01-Tags-Improvement.md
Comment (handrews) [reply proposals/2024-09-01-Tags-Improvement.md:128]: @ralfhandl I see- that makes sense.

File: proposals/2024-09-01-Tags-Improvement.md
Comment (lornajane) [reply proposals/2024-09-01-Tags-Improvement.md:13]: Ah, no I wanted to link something else here. I'll fix it

File: proposals/2024-09-01-Tags-Improvement.md
Comment (lornajane) [reply proposals/2024-09-01-Tags-Improvement.md:30]: It's a very common pattern of what downstream tools expect

File: proposals/2024-09-01-Tags-Improvement.md
Comment (lornajane) [reply proposals/2024-09-01-Tags-Improvement.md:48]: I don't love or hate any of the options

Comment (ralfhandl) [reply proposals/2024-09-01-Tags-Improvement.md:48]: @handrews It's probably only my experience with "Foo category" being a coarse-grained grouping of Foos, and "Foo type" being a refinement of "Foo category" 😄

**Review by ralfhandl — APPROVED** [review]

File: proposals/2024-09-01-Tags-Improvement.md
Comment (ralfhandl) [reply proposals/2024-09-01-Tags-Improvement.md:30]: Well...

**Review by kostiahuryuk — APPROVED** [review]

**Review by handrews — APPROVED** [review]
I'm approving this in its current form because I'd be OK with `kind` and, as noted in today's TDC call, this is just the proposal and we can continue to discuss the field name as it moves to reality.

More importantly, while I'm open to alternative proposals if someone comes up with one, I emphatically support adding the simple parent hierarchy.  In the Moonwalk discussion I advocated for a more thorough and flexible relational system, but I was convinced (and remain convinced) that a simple hierarchy is what is most requested, and is something that has more-or-less been implemented as extensions already.

So while I'd personally be inclined to a more complex knowledge management approach, I think this is the right step in the spirit of _incremental progress_ that really should define the 3.x line.

**Comment by baywet** [comment]
I'll try to recap in a more intelligible way what I mentioned.

What we have today are effectively labels/a list of terms. They only allow for a simple (one level/node) grouping of of operations and carry very little structure, let alone semantic meaning.

This proposal effectively "upgrades" things to a classification / hierarchy. It's really nice because:

- It maintains compatibility with existing investments.
- It's simple to understand.

We probably don't want to required anybody working with OpenAPI to have a background in information management, although that might bring interoperability with tooling from that industry. Enabling a more semantic classification of the operation could also better enable "autonomous agents" (i.e. AI orchestrators) to discover the operation.

I believe we should at least consider additional scenarios: matrix organizations. Think of the traditional insurance company that can offer packages for both auto and housing. Let's say they have dedicated quotes endpoints for each. Their taxonomy might look something like that. Each of the quotes operations being tied to its respective tag.

```
[root]
|---->Housing--->Housing_Prospects
|---->Auto-->Auto_Prospects
```

But there's no reason why I shouldn't be able to discover that Housing_propects and Auto_propects are related mechanically, and without having to require an LLM.

Hence why I'm suggesting that _in addition to_ the parent property, we should also at the very least introduce a "relatesTo" property (multi-value, one way).

An very short [article on the topic](https://www.copyright.com/blog/taxonomy-vs-ontology/)

Thoughts?

**Comment by ralfhandl** [comment]
@baywet Could you please formulate your addition in a new PR?

I'm merging this now as everyone seems to be ok with it.

**Pull request merged**
